persist/fs/write.go

```diff
@@ -164,7 +170,17 @@ func (w *writer) Write(key string, data []byte) error {
 	return w.WriteAll(key, [][]byte{data})
 }
 
-func (w *writer) WriteAll(key string, data [][]byte) error {
+func (w *writer) WriteAll(key string, data [][]byte) (retErr error) {
+	if w.err != nil {
+		return w.err
+	}
+
+	defer func() {
+		if retErr != nil {
```

Comment:
Defer is quite costly, also you never actually set `retErr` in this method so I believe it'll never get set?  It is my understanding unless you do a "naked" return, i.e. just `return` it'll take the value of what you return, not the value of `retErr`?